ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/state/PendingBlocksManager.java

```diff
@@ -127,6 +131,22 @@ public Optional<BlockHeader> lowestAnnouncedBlock() {
         .min(Comparator.comparing(BlockHeader::getNumber));
   }
 
+  /**
+   * Get the lowest pending ancestor block saved for a block
+   *
+   * @param block target block
+   * @return An optional with the lowest ancestor pending block
+   */
+  public Optional<Block> pendingAncestorBlockOf(final Block block) {
+    if (block == null) {
+      return Optional.empty();
+    }
+    if (pendingBlocks.containsKey(block.getHeader().getParentHash())) {
+      return pendingAncestorBlockOf(pendingBlocks.get(block.getHeader().getParentHash()).block());
+    }
```

Comment:
It's probably not a great idea to use a recursive algorithm here.  An iterative approach should be just as simple, likely more performant and not risk stack overflow.

Reply:
It makes sense. Done